nncf/experimental/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,679 @@
+
+
+IGNORED_NAME_OPERATORS = [*PTDropoutMetatype.get_all_aliases(), MODEL_OUTPUT_OP_NAME]
+
+
+class SearchGraphNode:
+
+    """
+    Class describing nodes used in SearchGraph.
+    """
+
+    def __init__(self,
+                 node_key: str,
+                 data: dict = None):
+        self.node_key = node_key
+        self.data = data if data else {}
+
+    @property
+    def is_merged(self) -> bool:
+        """
+        Returns value that the node is merged.
+        """
+        return self.data.get(SearchGraph.IS_MERGED_NODE_ATTR)
+
+    @property
+    def is_dummy(self) -> bool:
+        """
+        Returns value that the node is dummy.
+        """
+        return self.data.get(SearchGraph.IS_DUMMY_NODE_ATTR)
+
+    @property
+    def node_name(self) -> NNCFNodeName:
+        return self.data.get(NNCFGraph.NODE_NAME_ATTR)
+
+    @property
+    def node_type(self) -> str:
+        """
+        Returns type of node.
+        """
+        return self.data.get(NNCFGraph.NODE_TYPE_ATTR)
+
+    @property
+    def layer_name(self) -> str:
+        """
+        Returns the name of the layer to which the node corresponds.
+        """
+        return self.data.get(NNCFGraph.LAYER_NAME_ATTR)
+
+    @property
+    def main_id(self) -> int:
+        """
+        Returns the id of node. In case if the node is merged returns id of the first node from merged node list.
+        """
+        # return the first id node
+        if not self.is_merged:
+            return self.data.get('id')
+        return self.data.get(SearchGraph.MERGED_NODES_NODE_ATTR)[0].get('id')
+
+    @property
+    def bottom_id(self) -> int:
+        """
+        Returns the id of node. In case if the node is merged returns id of the last node from merged node list.
+        """
+        if not self.is_merged:
+            return self.data.get('id')
+        return self.data.get(SearchGraph.MERGED_NODES_NODE_ATTR)[-1].get('id')
+
+    def __str__(self):
+        return self.node_key
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __eq__(self, other):
+        return isinstance(other, SearchGraphNode) and self.node_key == other.node_key
+
+
+class BuildingBlock:
+
+    """
+    Describes a building block that is uniquely defined by the start and end nodes.
+    """
+
+    def __init__(self, start_node: SearchGraphNode, end_node: SearchGraphNode):
+        self.start_node = start_node
+        self.end_node = end_node
+
+
+class BuildingBlockType(Enum):
+    """
+    Describes type of building block for transformers-based network.
+    `MSHA` type is characterized by the presence 4 FC and 2 MatMul layers.
+    `FF` type is characterized by the presence 2 FC layers.
+    """
+    MSHA = 'MSHA'
+    FF = 'FF'
+    Unknown = 'unknown'
+
+
+class BuildingBlockInfo:
+
+    """
+    Describes additional information about the building block
+    the address of each layer, the modules contained and type of block(optional).
```

Comment:
If block type is optional, why is the corresponding init param not optional?

Reply:
Fixed.